pkg/diff/sql_generator.go

```diff
@@ -1506,26 +1613,108 @@ func (csg *checkConstraintSQLGenerator) Delete(con schema.CheckConstraint) ([]St
 	}}, nil
 }
 
-func (csg *checkConstraintSQLGenerator) Alter(diff checkConstraintDiff) ([]Statement, error) {
-	if cmp.Equal(diff.old, diff.new) {
-		return nil, nil
+func (csg *checkConstraintSQLVertexGenerator) Alter(diff checkConstraintDiff) ([]Statement, error) {
+	oldCopy := diff.old
+
+	var stmts []Statement
+	if !diff.old.IsValid && diff.new.IsValid {
+		stmts = append(stmts, validateConstraintStatement(csg.tableName, schema.EscapeIdentifier(diff.new.Name)))
+		oldCopy.IsValid = diff.new.IsValid
 	}
 
-	oldCopy := diff.old
-	oldCopy.IsValid = diff.new.IsValid
-	if !cmp.Equal(oldCopy, diff.new) {
+	// Normalize the key columns, since order does not matter.
+	sort.Strings(oldCopy.KeyColumns)
+	newCopy := diff.new
+	sort.Strings(newCopy.KeyColumns)
+	if !cmp.Equal(oldCopy, newCopy) {
 		// Technically, we could support altering expression, but I don't see the use case for it. it would require more test
 		// cases than force re-adding it, and I'm not convinced it unlocks any functionality
 		return nil, fmt.Errorf("altering check constraint to resolve the following diff %s: %w", cmp.Diff(oldCopy, diff.new), ErrNotImplemented)
-	} else if diff.old.IsValid && !diff.new.IsValid {
-		return nil, fmt.Errorf("check constraint can't go from invalid to valid")
 	} else if len(diff.old.DependsOnFunctions) > 0 || len(diff.new.DependsOnFunctions) > 0 {
 		return nil, fmt.Errorf("check constraints that depend on UDFs: %w", ErrNotImplemented)
 	}
 
-	return []Statement{
-		validateConstraintStatement(csg.tableName, schema.EscapeIdentifier(diff.new.Name)),
-	}, nil
+	return stmts, nil
+}
+
+func (*checkConstraintSQLVertexGenerator) GetSQLVertexId(con schema.CheckConstraint) string {
+	return buildVertexId("checkconstraint", con.Name)
+}
+
+func (csg *checkConstraintSQLVertexGenerator) GetAddAlterDependencies(con, _ schema.CheckConstraint) ([]dependency, error) {
+	deps := []dependency{
+		mustRun(csg.GetSQLVertexId(con), diffTypeDelete).before(csg.GetSQLVertexId(con), diffTypeAddAlter),
+	}
+
+	targetColumns, err := getTargetColumns(con, csg.newSchemaColumnsByName)
+	if err != nil {
+		return nil, fmt.Errorf("getting target columns: %w", err)
+	}
+
+	isOnPreExistingColumn := false
+	if len(targetColumns) == 1 {
+		targetColumn := targetColumns[0]
+		if _, ok := csg.addedColumnsByName[targetColumn.Name]; !ok {
+			isOnPreExistingColumn = true
+		}
+	}
+
+	if isOnPreExistingColumn && isValidNotNullCC(con) {
+		// If the NOT NULL check constraint is on a pre-existing column, then we should ensure it is added before
+		// the column.
```

Comment:
Is this correction right?
```suggestion
		// the column alter.
```